SimCalorimetry/EcalSimAlgos/src/EcalLiteDTUCoder.cc

```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+  // fill ADC trace in gain 0 (x10) and gain 1 (x1)
+  for (unsigned int igain = 0; igain < ecalPh2::NGAINS; ++igain) {
+    for (int i(0); i != nSamples; ++i) {
+      double asignal = 0;
+      if (!m_PreMix1) {
+        asignal = pedestals[igain] + ecalSamples[i] / (LSB[igain] * icalconst) + trueRMS[igain] * noiseframe[igain][i];
+        //Analog signal value for each sample in ADC.
+        //It is corrected by the intercalibration constants
+
+      } else {
+        //  no noise nor pedestal when premixing
+        asignal = ecalSamples[i] / (LSB[igain] * icalconst);
+      }
+      int isignal = asignal;
+
+      unsigned int adc = asignal - (double)isignal < 0.5 ? isignal : isignal + 1;
+      // gain 0 (x10) channel is saturated, readout will use gain 1 (x10), but I count the number of saturated samples
+      if (adc > ecalPh2::MAXADC) {
+        adc = ecalPh2::MAXADC;
+        if (nSaturatedSamples == 0)
+          firstSaturatedSample[igain] = i;
+        nSaturatedSamples++;
+      }
+      adctrace[i][igain] = adc;
+    }
+    if (nSaturatedSamples == 0) {
+      break;  //  gain 0 (x10) is not saturated, so don't bother with gain 1
+    }
+  }  // for igain
+
+  int igain = 0;
+
+  //Limits of gain 1:
+  //The Lite DTU sends 5 samples before the saturating one, and 10 after with gain 1.
+  //we put the maximum in bin 5, but could happen that the system saturates before.
+
+  int previousSaturatedSamples = 5;
+  int nextSaturatedSamples = 10;
+  int startingLowerGainSample = 0;
+  int endingLowerGainSample = (firstSaturatedSample[0] + nextSaturatedSamples + (nSaturatedSamples));
+
+  if (nSaturatedSamples != 0 and (firstSaturatedSample[0] - previousSaturatedSamples) < 0) {
+    startingLowerGainSample = 0;
+  } else {
+    startingLowerGainSample = (firstSaturatedSample[0] - previousSaturatedSamples);
+  }
+
+  //Setting values to the samples:
+  for (int j = 0; j < nSamples; ++j) {
+    if (nSaturatedSamples != 0 and j >= startingLowerGainSample and j < endingLowerGainSample) {
+      igain = 1;
+    } else {
+      igain = 0;
+    }
+    df.setSample(j, EcalLiteDTUSample(adctrace[j][igain], igain));
+  }
+}
+
+void EcalLiteDTUCoder::findPedestal(const DetId& detId, int gainId, double& ped, double& width) const {
+  EcalLiteDTUPedestalsMap::const_iterator itped = m_peds->getMap().find(detId);
+  ped = (*itped).mean(gainId);
+  width = (*itped).rms(gainId);
+
+  LogDebug("EcalLiteDTUCoder") << "Pedestals for " << detId.rawId() << " gain range " << gainId << " : \n"
+                               << "Mean = " << ped << " rms = " << width;
+}
+
+void EcalLiteDTUCoder::findIntercalibConstant(const DetId& detId, double& icalconst) const {
+  EcalIntercalibConstantMC thisconst = 1.;
+  // find intercalib constant for this xtal
+  const EcalIntercalibConstantMCMap& icalMap = m_intercals->getMap();
+  EcalIntercalibConstantMCMap::const_iterator icalit = icalMap.find(detId);
+
+  thisconst = (*icalit);
```

Comment:
this may not be safe, because `find()` above could return an iterator to the end of the container. check for that explicitly before dereferencing the iterator